### PR TITLE
Try reducing specificity of layout margin rules

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

Hypothetical alternative to #47399: trying to reduce the specificity of the layout rules to the point that they don't override block global margins.

This assumes that in a layout container:

- first child top margin is always 0;
- last child bottom margin is always 0;
- top margins of non-first children are always set to gap value and bottom margins to 0, UNLESS the child itself has its own margin value.

## Why?

#43404, #44736, #33795, #40813, #35267 and #46902... the issues keep piling up. Essentially, layout spacing styles shouldn't override global block margins.

## Testing Instructions

See #47399.

### CODE REVIEW

**Comment by tellthemachines** [comment]
> It seems to be working! except for the small issue of global spacing values in Group block being overridden by the default because they now both have 0 specificity 😂

Update: I bumped the specificity of the block rule ever so slightly by prefacing it with `:root`, and now it's working as expected:
Screenshot 2023-02-09 at 12 04 43 pm

**Review by andrewserong — COMMENTED** [review]
This is a very compelling exploration! Really cool that it appears to largely resolve the block-level-in-global styles margin case 👍

One challenge here, though, is the introduction of `!important` rules. From a quick test, it looks like the layout's `!important` rules for the first and last child blocks results in those rules overriding block-level-in-block-attributes margin rules:

image

In the above screenshot, this particular cover block has a `margin-top` value set to one of the spacing presets, and that style is output in the block's inline styles. However, the layout support's rule wins out:

image

**Comment by andrewserong** [comment]
Just noticed another couple of challenges with reducing the layout specificity / removing the bottom margin overrides and existing block styles:

* The Cover block includes a margin bottom rule of `1.75em` ([here](https://github.com/WordPress/gutenberg/blob/d9cb56caff47c320bb6efff1c50f541aa0cb45cc/packages/block-library/src/columns/style.scss#L3)) which doesn't get overridden by the layout rules in this PR (due to the removal of resetting the bottom margin + lower rules specificity?)

image

* For themes that opt in to the `wp-block-styles` theme support (e.g. TT2), then the audio block outputs [a figure element margin override](https://github.com/WordPress/gutenberg/blob/fd3b08bef5da71ca02db872ddd57a7a77beaec21/packages/block-library/src/audio/theme.scss#L5), which wins out over the layout rules in this PR.

image

So, in terms of evaluating whether we can reliably reduce the layout specificity, a couple of questions might be:

* If we were to audit the core blocks and remove all margin rules (or reduce specificity) that would conflict with layout, would that be reliable enough to allow the layout rules to be lowered in specificity?
* Do 3rd party blocks set margin rules and expect layout rules (E.g. Group block) to overwrite those margin rules? I suspect this might be the case, given that core blocks already have margin rules that layout overrides 🤔

In any case, I think the ideas in this PR are super valuable to explore, particularly given that the complexity of the change (and style output) is much simpler than what's being explored over in https://github.com/WordPress/gutenberg/pull/47399. Thanks again for digging in here!

**Comment by tellthemachines** [comment]
Thanks for thoroughly poking holes in this experiment @andrewserong 😄 

I removed the `!important`s and managed to fiddle with `:where()` and pseudoclass repetition to the point that I think everything is working in the correct order now. (The pseudoclass repetition was only necessary because otherwise in the editor there wasn't enough specificity to override global margins applied on a class selector such as `.wp-block-heading`; it actually worked with a single one in the front end.)

> * If we were to audit the core blocks and remove all margin rules (or reduce specificity) that would conflict with layout, would that be reliable enough to allow the layout rules to be lowered in specificity?
> * Do 3rd party blocks set margin rules and expect layout rules (E.g. Group block) to overwrite those margin rules? I suspect this might be the case, given that core blocks already have margin rules that layout overrides 🤔

These are great questions! 

Regarding 1, I can go through the core blocks in this PR and try to lower specificity where needed. I'm pretty sure we can manage that.

As for 2, it's hard to know what might be expected, but over time we've consistently had pointed complaints about the over-specificity of our styles, and I suspect we won't get much pushback on lowering them. It is a breaking change, and should be announced as such (dev notes etc.) but ultimately any change in specificity is a breaking change, and even just adding rules can break something somewhere. I'm optimistic that we can get away with it 😅

**Review by andrewserong — COMMENTED** [review]
Thanks for continuing down this particular path @tellthemachines, I still think it's a really compelling direction 👍

From giving this a re-test, I think there's another wrinkle with the update to the non-first-or-last rules, in that it looks like we still need to reset the `margin-block-end` for each of these rules to `0` so that we can override the user agent styles for elements like paragraph tags. Otherwise, it looks like the `1em` rules creep back into play. Here's a screenshot:

image

I'd be curious to hear what other folks think about this as a potential solution, so I'll just copy in @t-hamano for visibility, too.

**Comment by tellthemachines** [comment]
I re-added the bottom margin override and reduced the specificity of the Columns block margin rule. I _think_ everything's working as it should now. 

Regarding a possible clash with #47952, I'm pretty sure we can make things work with the compound class added in that PR, because the rule as it is in this PR - `:root :where(%s.%s) %s` only needs the specificity of one classname, and that could be the compound classname instead of `:root`.

I'm going to go ahead and mark this ready for review, but I'll hold off on updating the test markup for now because it's likely that #47952 will be merged soon, and I don't love solving merge conflicts _that_ much 😂

**Comment by andrewserong** [comment]
Thanks for the update @tellthemachines! From a quick search of blocks' CSS, here are some other blocks where the specificity may need to be reduced (or at least, we should test to see whether or not it'll need adjusting)

* File: has bottom margin
* Image: left aligned top/bottom margins
* Image: right aligned top/bottom margins
* Post Excerpt block: looks like [top/bottom margin rules](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/post-excerpt/style.scss#L2) might have been added recently? (The ones at the root level of the block might be able to be removed?)
* Post Template block (this one zeros out top/bottom margins, so might be able to be parked until we do the post template block refactor, but we should probably test it anyway)
* Pullquote block: has shorthand margin rule

**Comment by MaggieCabrera** [comment]
Maybe [this](https://github.com/WordPress/gutenberg/pull/43905#issuecomment-1455797661) is another example of this being the case?

**Comment by tellthemachines** [comment]
Thanks for the list of blocks with margins @andrewserong ! I lowered the specificity on File, Post Excerpt and Pullquote. Left and right-aligned images only seem to have inline margins, unless I'm missing something 🤔 and in my testing Post Template seems to be getting the right styles whichever position it's in inside Query.

Regarding #45107 and similar uses of `"__experimentalStyle"`: is there anything stopping us from wrapping the output of `"__experimentalStyle"` in a `:where`? Any margins added to blocks in that way will override the layout styles in this PR.

**Comment by andrewserong** [comment]
> is there anything stopping us from wrapping the output of "__experimentalStyle" in a :where? Any margins added to blocks in that way will override the layout styles in this PR.

I don't think so. It would probably involve adding a bit of logic — currently I _think_ `__experimentalStyle` is implemented in a fairly naive way in that the values entered there wind up being part of the theme.json structure, as if the theme had provided those values at the block level. Separately to this PR, I think it'd be good to explore your idea of ensuring that layer of styles (those provided by the block) have a lowered specificity in a `:where()` rule — or, at the very least, that margin rules have it 👍

**Comment by andrewserong** [comment]
> Left and right-aligned images only seem to have inline margins, unless I'm missing something 🤔

I did a little bit of digging: it looks like there are top/bottom margin rules for a deprecated form of the image block that target `.alignleft` and `.alignright` but only as a child of the image block's wrapper ([here](https://github.com/WordPress/gutenberg/blob/21da6b3f53abd093dd494afb9757ec69a8d72bcc/packages/block-library/src/image/style.scss#L57) and [here](https://github.com/WordPress/gutenberg/blob/21da6b3f53abd093dd494afb9757ec69a8d72bcc/packages/block-library/src/image/style.scss#L67)), so they won't interfere with these layout rules changes. As far as I can tell, those alignment rules still exist to support unmigrated old image block markup. The PR that removed the wrapper div (where these rules used to apply) was https://github.com/WordPress/gutenberg/pull/38657. So, looks like we're all good for the image block, at least!

**Comment by tellthemachines** [comment]
Now that #47952 is merged, I've updated the rules in this PR to take those changes into account. I also updated the global styles output to reflect the changed rules and fixed the test markup. This is now ready for another round of review/testing!

**Review by andrewserong — COMMENTED** [review]
This is testing really well for me so far @tellthemachines, and I think it's looking like the most promising direction for addressing the styling conflicts between margin and layout rules 👍

Just adding a few more pings for visibility now that this is ready for another round of review, as some of these folks have been involved in the discussions so far: @t-hamano, @aaronrobertshaw, @youknowriad, @WordPress/gutenberg-design 

Will do some more detailed testing when I'm back on Tuesday!

**Comment by jasmussen** [comment]
I'd love a @youknowriad take on this.

In principle, I'm a fan of using `where`, however we've seen some pushback on the usage for it for functionality that's crucial for basic functionality, such as navigation menu dropdowns open/close behaviors. While the use of legacy browsers will drop every single day to eventually become negligible, it's worth considering the feedback in light of margins. Does the benefit of vastly reduced specificity outweigh the downsides? _Probably_, yes, as noted it wasn't the case for navigation dropdowns.

I haven't had a chance to deeply test this either, but from a glance it looks like it's changing the owl selector and the "stack" behavior. Is the net result the same? That is: it's so nice to not have to worry about the top margin of the first child, and the bottom margin of the last child, and have the margin behave like "gap", essentially. Ideally we can keep that behavior.

Adjacent, I looked at using `margin-inline` as a replacement for centering a column, and still allowing negative margins left and right: https://codepen.io/joen/pen/vYrYqya — useful if we want left/right margin controls to be able to be negative. Doesn't seem entirely related as that property is mainly for the horizontal positioning, but wanted to share in case it inspired ideas.

**Comment by youknowriad** [comment]
To be honest, I've lost track of all the layout related changes that I'm not really able to give an informed opinion anymore. I feels like we need some architecture documentation about this specific feature to help with future maintainability. 

By looking quickly at the PR, the thing that stood out to me was the changes to individual block styles and was wondering how much this could impact third party blocks and forcing them to update their own styles.

**Comment by tellthemachines** [comment]
> In principle, I'm a fan of using `where`, however we've seen some pushback on the usage for it for functionality that's crucial for basic functionality, such as navigation menu dropdowns open/close behaviors. While the use of legacy browsers will drop every single day to eventually become negligible, it's worth considering the feedback in light of margins. Does the benefit of vastly reduced specificity outweigh the downsides? _Probably_, yes, as noted it wasn't the case for navigation dropdowns.

True, I'd be concerned if the styles were essential for readability or accessibility, but in this case (as with similar places we're using `where`) the absence of these styles will only make the page a little uglier, not unusable. The only major browser not supporting it is IE, and IE users at this point will be very used to websites not looking their best 😅 

> I haven't had a chance to deeply test this either, but from a glance it looks like it's changing the owl selector and the "stack" behavior. Is the net result the same? That is: it's so nice to not have to worry about the top margin of the first child, and the bottom margin of the last child, and have the margin behave like "gap", essentially. Ideally we can keep that behavior.

Yes, it should be the same, and as far as my testing goes it's working well. The main point of difference is we want the spacing values on blocks that _aren't_ the first or last child to not override any global margin values the blocks might have, so the generic rule for all blocks had to have much lower specificity than the zero top margin for first child, and zero bottom margin for last child.

To make this work, it was necessary to also reduce the specificity of margins set on some core blocks, because we still want the layout spacing to override them. To Riad's point:

> By looking quickly at the PR, the thing that stood out to me was the changes to individual block styles and was wondering how much this could impact third party blocks and forcing them to update their own styles.

yes, it's possible that third party blocks with margin styles that before were overridden by layout styles might now instead see their margins overriding the layout ones, but that seems to me a lesser consequence than the potential impact of _increasing_ specificity on global block styles, or adding complex logic to generate a bunch of new rules so global block margins can override layout styles. 

What are everyone else's thoughts on this?

> I feels like we need some architecture documentation about this specific feature to help with future maintainability.

Yes, this would be very useful. It took me ages to wrap my head around layout logic when I started work on it a year or so ago, and it's grown a fair bit since then 😅

**Comment by andrewserong** [comment]
> I feels like we need some architecture documentation about this specific feature to help with future maintainability.

We have a layout section in the styles architecture page (https://developer.wordpress.org/block-editor/explanations/architecture/styles/#layout-styles), but yes, it'd be good to expand on this and possibly give it its own page for the next release, along with stabilising the block support.

**Comment by andrewserong** [comment]
> yes, it's possible that third party blocks with margin styles that before were overridden by layout styles might now instead see their margins overriding the layout ones, but that seems to me a lesser consequence than the potential impact of increasing specificity on global block styles, or adding complex logic to generate a bunch of new rules so global block margins can override layout styles.

> What are everyone else's thoughts on this?

It's a difficult trade-off, as either way it involves a potentially breaking change, but after exploring the alternative over in #47399 of adding in additional margin rules with ever increasing complexity, I quite prefer the ideas proposed in this PR. The approach here feels a little more "correct" to me, in that its goal is to create a base set of rules that can be more easily overridden. Another concern I had with #47399 was that the additional complexity and rules are then harder to maintain or for other contributors to jump in and work with.

I suppose when it comes to 3rd party blocks, just as with core blocks, one of the challenges is that blocks might ship with margin values so that the spacing works nicely in classic themes, so there is the potential there for those blocks to no longer have pleasing spacing in blocks themes. On the other hand, the theme developer or site owner still has control over being able to set block spacing at the parent level of a container to override those margins, so a site could still have a mechanism to override spacing as they desire.

In short, I think the ideas in this PR ultimately allow a site owner to have greater flexibility and control over their site's spacing, so it still seems like a promising direction to me 🙂

**Comment by aaronrobertshaw** [comment]
> yes, it's possible that third party blocks with margin styles that before were overridden by layout styles might now instead see their margins overriding the layout ones

I haven't yet caught up on the latest discussions here but there's another angle to view this from. Layout support styles introduced a regression where they overrode existing margin styles (block support or otherwise). For blocks impacted by this, the restoration of their margin styles would be a bug fix.

> In short, I think the ideas in this PR ultimately allow a site owner to have greater flexibility and control over their site's spacing, so it still seems like a promising direction to me 🙂

Sounds like a win to me.

**Comment by tellthemachines** [comment]
Hey folks! I've addressed all pending feedback so I guess this is ready for another round of reviews? Would be good to move forward as there are quite a few pending issues caused by over-specific layout styles 😅

**Comment by tellthemachines** [comment]
Adding @WordPress/gutenberg-core and @WordPress/gutenberg-design to see if we can get this moving. All testing, reviews, questions and concerns welcome!

**Review by andrewserong — COMMENTED** [review]
I think these changes to the layout rules are the best idea yet to make layout rules play nicely with margin rules, and I also like that it reduces their specificity overall 👍

In re-testing, I noticed that this PR appears to override the bottom margin of the post title in the post editor. With TwentyTwentyThree active, the `.editor-styles-wrapper :where(body .is-layout-flow) > :last-child:last-child {` rule overrides `.editor-styles-wrapper .wp-block-post-title {`. This appears to be because of the `is-layout-flow` class being added to the post title wrapper and the increased specificity of the first/last layout rules:

| Trunk | This PR |
| --- | --- |
| image | image |

To fix this, I imagine we could remove the `is-layout-flow` on the `edit-post-visual-editor__post-title-wrapper`? Alternately, will we run into other issues now that the first/last positions have greater specificity than on trunk?

**Review by andrewserong — COMMENTED** [review]
Just did a pretty thorough test of inserting all the core blocks rendered by the Style Book, and couldn't find any issues with margins that aren't already present on trunk (e.g. very subtle things like the caption of embed and video blocks having a negative margin). So they're not a blocker for this PR 👍

Also checked:

✅ Tested blocks in both the post and site editors
✅ Columns block updated margin plays nicely with layout rules
✅ File block updated margin plays nicely with layout rules
✅ Post Excerpt block updated margin plays nicely with layout rules
✅ Pullquote block updated margin plays nicely with layout rules

Once the issue with the post title within the post editor is fixed up, I think this PR would be worth trying! Would be keen to hear what other folks think, too 🙂

**Comment by tellthemachines** [comment]
> To fix this, I imagine we could remove the is-layout-flow on the edit-post-visual-editor__post-title-wrapper? 

Thanks for testing and finding this bug! Yeah I think here it makes the most sense to remove `is-layout-flow` from the post title wrapper. The only thing that layout class does is add spacing rules between blocks, and given that wrapper only ever has one block and we're defining margins for it already, the class is useless. It also does nothing on classic themes, where the spacing rules for flow layout aren't even output at all.

> Alternately, will we run into other issues now that the first/last positions have greater specificity than on trunk?

It depends on what our intended outcome is. If we want first/last children to _always_ have 0 top/bottom margin, these styles assure that will happen. The difference in specificity for those styles between trunk and this branch is 0-1-1 vs 0-2-0, so they are slightly more specific. However, the effect on blocks with global margins will be the same as on trunk. The only situation I can think of where this change could make a difference is in a third-party block that has custom margin styles with a specificity bigger than 0-1-1 but smaller than (or equal to) 0-2-0.

**Comment by andrewserong** [comment]
> Yeah I think here it makes the most sense to remove is-layout-flow from the post title wrapper.

Sounds good 👍 

> It depends on what our intended outcome is. If we want first/last children to always have 0 top/bottom margin, these styles assure that will happen.

Yes, I think so far that's been our assumption for the layout support, and I think the change here sounds reasonable, and it doesn't cause any issues with core blocks. Also, while the change here is impactful, it's fairly small in terms of code changes, so it'd be a fairly quick revert if we need to roll back.

I can give this a final review again on Monday if no-one beats me to it, and I reckon this PR would be good to try out in trunk!

Just a note that the unit tests are currently failing due to an issue on trunk. Once https://github.com/WordPress/gutenberg/pull/49497 is in, and this PR is rebased again, they should be fixed up 🙂

**Comment by tellthemachines** [comment]
> Just a note that the unit tests are currently failing due to an issue on trunk. Once https://github.com/WordPress/gutenberg/pull/49497 is in, and this PR is rebased again, they should be fixed up 🙂

Thanks, I was wondering about that 😅

**Comment by andrewserong** [comment]
> Thanks, I was wondering about that 😅

Just unfortunate rebase timing! 😄

File: lib/theme.json [hunk 1]
```diff
@@ -341,7 +352,7 @@
 							}
 						},
 						{
-							"selector": " > *",
+							"selector": " > :where(*)",
```
Comment (andrewserong) [inline lib/theme.json]: After the rebase, do we need to update the `grid` layout type to also use this specificity?

**Review by andrewserong — COMMENTED** [review]
This is looking _so_ close to land to me. After a re-rest, I realised that previously I was always testing in the site editor within a Group block, but not at the root of the document. It looks like we might need to update the `.wp-site-blocks` rules to use the same specificity as the layout types in `theme.json`?

E.g. here: https://github.com/WordPress/gutenberg/blob/4734fb62f3f879757ed69d6559ec3c18c1898e4c/lib/class-wp-theme-json-gutenberg.php#L2491

And here:
https://github.com/WordPress/gutenberg/blob/163185184e80f2a8d0cb71a69b5844849d0da542/packages/block-editor/src/components/global-styles/use-global-styles-output.js#L949

Without it, when adding blocks at the root of the document, the block-level margin rules in global styles don't take precedence. Here's a quick screengrab testing with the Cover block:

https://user-images.githubusercontent.com/14988353/229391071-f1b96848-949e-4cbe-9d50-43dd3dfccb30.mp4

**Comment by tellthemachines** [comment]
Thanks for restesting @andrewserong and yes to both questions 😄

File: lib/theme.json (hunk 1 above)
Comment (tellthemachines) [reply lib/theme.json]: Actually, looking better at it, we don't! the reason being the `*` selector has 0 specificity already. I'm not sure why I wrapped it in a `:where` initially 🤔 . Will update for consistency!

File: lib/class-wp-theme-json-gutenberg.php [hunk 2]
```diff
@@ -2488,8 +2488,9 @@ public function get_root_layout_rules( $selector, $block_metadata ) {
 		$has_block_gap_support = _wp_array_get( $this->theme_json, array( 'settings', 'spacing', 'blockGap' ) ) !== null;
 		if ( $has_block_gap_support ) {
 			$block_gap_value = static::get_property_value( $this->theme_json, array( 'styles', 'spacing', 'blockGap' ) );
-			$css            .= '.wp-site-blocks > * { margin-block-start: 0; margin-block-end: 0; }';
-			$css            .= ".wp-site-blocks > * + * { margin-block-start: $block_gap_value; }";
+			$css            .= ":where(.wp-site-blocks) > * { margin-block-start: $block_gap_value; }";
```
Comment (andrewserong) [inline lib/class-wp-theme-json-gutenberg.php]: It looks like this and the JS version are missing the `margin-block-end: 0;` rule, so the user agent stylesheet rules for paragraph are poking through:

image

Comment (tellthemachines) [reply lib/class-wp-theme-json-gutenberg.php]: Oh whoops, I copied the old rule and forgot to add the end margin 😳

Comment (tellthemachines) [reply lib/class-wp-theme-json-gutenberg.php]: Fixed!

**Review by andrewserong — APPROVED** [review]
Nice one, thanks for all the updates @tellthemachines and for persevering with this! All the issues I've raised are now resolved, and this is testing well:

✅ Re-tested with a wide range of blocks, and couldn't find any issues
✅ Spacing between post title and post content within the post editor now respects the margin rules for the post title block
✅ Root-level site block spacing works as expected, with rules correctly overridden by block margin rules where appropriate
✅ Block-level margin rules in global styles apply as expected in positions within the layout container
✅ Global styles rules are overridden correctly by block-level styles

I think this is in a great place to land now, and would be worth trying out in the plugin.
LGTM! ✨

**Comment by tellthemachines** [comment]
## Dev note

### Changes in CSS specificity for some layout types

The CSS output for the `margin` styles of children of `constrained` and `flow`(default) layout containers has changed. The generic rules used to have a specificity of 0,1,0 for all blocks; they have now changed to 0,0,0, for all blocks with an extra 0,2,0 rule applied only to the first and last blocks in the container.

This fixes the issue of global margin styles for specific blocks being overridden by layout styles (see https://github.com/WordPress/gutenberg/issues/43404).
